### ISSUE TITLE
Can't set dynamic route requirements in an extension

### ISSUE THREAD

**Issue by mattvick** [bolt/bolt#3366]
Can't add a dynamic route requirement to a Bolt route.

I'm defining my route as follows:

```yaml
formpages:
    path: '/{contenttypeslug}/{slug}'
    defaults:
        _controller: 'Bolt\Extension\Mahango\Forms\Extension::record'
    requirements:
        contenttypeslug: 'page'
        slug: 'Bolt\Extension\Mahango\Forms\Extension\Routing::getSlugRequirement'
    method: 'GET|POST'
```

This needs to be defined in `routing.yml` because if I use `$app->match()` in my extension the route is added with the lowest priority and I need it to be above the core `contentlink` route.

In my `getSlugRequirement()` method I want to query the database to get a subset of page records but I don't have access to the `$app` var when the method is being called.

I've looked at the core code and tried to use `$app->mount('', new Bolt\Extension\Mahango\Forms\Extension\Routing());` in my `bootstrap.php` but here the extension hasn't yet been initialised. It seems like whichever angle I come at this I don't ave access to the other side.

Is this a limitation of extensions?

How can I do get this to work?

**Comment by rixbeck** [comment]
You should implement your own router classes without any problems in extensions as well, like I do [here](https://github.com/rixbeck/bolt-extension-lang/blob/master/src/Controller/LanguageController.php).
You can overwrite default `contentlink` routes with your controller method in a class that extends original `Bolt\Controllers\Frontend` and calls the appropriate parent method, after you handle the request by that way you need. (...in the case I interpreted the problem correctly)

**Comment by GawainLynch** [comment]
Routing priorities are being addressed in 2.3 and as @rixbeck mentions, extending `Frontend` is your best best presently.
